File: AI_Files/00_player_example.py
#!/usr/bin/env python3

# How to have file run as an executable: chmod 775 FILENAME.py

# Grabbing some constants similar to defines.h
# Shot related
SHOT = '@'
MISS = 'O'
DUPLICATE_SHOT  = '!'
HIT  = 'X'
KILL = 'K'
OPPONENT_SHOT = '*'
SHIP = 'S'
WATER = '~'
INVALID_SHOT = '\0'
PLACE_SHIP = 'Z'
# enum Direction
NONE=0
HORIZONTAL=1
VERTICAL=2

import os
import socket
import sys
import json
import copy


gameVars = {
    "boardSize": 0,
    "shipBoard": [],
    "shotBoard": [],
}

def updateBoard(board, row, col, length, direction, newChar):
    if direction == HORIZONTAL:
        for len in range(length):
            gameVars[board][row][col+len] = newChar
    elif direction == VERTICAL:
        for len in range(length):
            gameVars[board][row+len][col] = newChar
    elif direction == NONE:
        gameVars[board][row][col] = newChar


def wipeBoards():
    array = []
    gameVars["shipBoard"]=[]
    gameVars["shotBoard"]=[]
    for _ in range(gameVars["boardSize"]):
        array.append(WATER)
    for _ in range(gameVars["boardSize"]):
        gameVars["shipBoard"].append(copy.deepcopy(array))
        gameVars["shotBoard"].append(copy.deepcopy(array))
    return


def sendGameVars(msg):
    gameVars["boardSize"] = int(msg["str"])
    msg["str"] = "Matthew Bouch"


def placeShip(msg): # This is an extremely bad-to-use algorithm in the long term--for more reasons than just the obvious one. Try writing your own from scratch.
    for row in range(gameVars["boardSize"]-msg["length"]):
        for col in range(gameVars["boardSize"]-msg["length"]):
            if(gameVars["shipBoard"][row][col] == WATER):
                msg["row"] = row
                msg["col"] = col
                msg["dir"] = HORIZONTAL
                for len in range(msg["length"]):
                    if(gameVars["shipBoard"][row][col] != WATER):
                        msg["dir"] = VERTICAL
                        updateBoard("shipBoard", row, col, msg["length"], VERTICAL, SHIP)
                        return
                updateBoard("shipBoard", row, col, msg["length"], HORIZONTAL, SHIP)
                return


def shootShot(msg):
    for row in range(0, gameVars["boardSize"]):
       for col in range(0, gameVars["boardSize"]):
           if(gameVars["shotBoard"][row][col]==WATER):
               msg["row"]=row
               msg["col"]=col
               updateBoard("shotBoard", row, col, 1, NONE, SHOT)
               return
        

def shotReturned(msg):
    # if(msg["client"]==clientID): # this is your AI's shot
        # do something
    # else: # this is your opponant's shot
        # do nothing
        # unless...?
    pass




def messageHandler(msg, clientID):
    if(msg["messageType"] == "setupGame"):
        msg["client"]=clientID
        sendGameVars(msg)
        # Populating boards with blank data
        wipeBoards()
    elif(msg["messageType"] == "matchOver"):
        msg["client"]=clientID
        wipeBoards()
    elif(msg["messageType"] == "placeShip"):
        msg["client"]=clientID
        placeShip(msg)
    elif(msg["messageType"] == "shootShot"):
        msg["client"]=clientID
        shootShot(msg)
    elif(msg["messageType"] == "shotReturn"):
        shotReturned(msg)
    elif(msg["messageType"] == "shipDied"):
        updateBoard("shipBoard", msg["row"], msg["col"], msg["length"], msg["dir"], KILL)
    elif(msg["messageType"] == "killedShip"):
        updateBoard("shotBoard", msg["row"], msg["col"], msg["length"], msg["dir"], KILL)






def main():
    global clientID
    
    # Create a UDS socket
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)

    # Connect the socket to the port where the server is listening
    server_address = './serversocket'
    # print('connecting to {}'.format(server_address))
    try:
        sock.connect(server_address)
    except socket.error as msg:
        print(msg)
        sys.exit(1)


    # Setup Game Vars
    clientID = os.getpid()
    gameVars["boardSize"] = 10

    # Populating boards with blank data
    #wipeBoards()

    while True:

        msg = json.loads(sock.recv(1500))
        # print("Received:", msg)
        
        messageHandler(msg, clientID)
        
        toSend = (json.dumps(msg, separators=(',', ':'))).encode()
        # print("Sending:", toSend)

        sock.send(toSend)


        #break
    sock.close()

if __name__ == "__main__":
    main()

File: AI_Files/00_player_example.cpp

#include "ai_functions.h"

void placeShip(json &msg);
void shootShot(json &msg);
void shotReturned(json &msg);

/**
 * @struct container
 * 
 * @brief container is used as a global variable for important items for you to keep track of.
 */
struct container{
    int  boardSize=10;
    char shipBoard[10][10];
    char shotBoard[10][10];
};
container gameVars;


/**
 * @brief Happens at the beginning of the battleships game only happens once (in 500 matches it only occurs once).
 * 
 * @param msg Used to give you the boardsize and for you to give your author name in response.
 *      json msg = {
 *          {"str", "10"}
 *      }
 */
void sendGameVars(json &msg){
    string temp = msg.at("str");
    gameVars.boardSize = stoi(temp);
    msg.at("str") = "Joey Gorski"; // Your author name(s) here
}

/**
 * @brief Parses the messageType, then calls the appropriate function. 
 * 
 * @param clientID The current AI's ID.
 * @param msg Contains which action to take.
 *      json msg = {
 *          {"messageType", "placeShip"}
 *      }
 */
void messageHandler(json &msg, string &clientID){
    if(msg.at("messageType")=="setupGame"){
        msg.at("client") = clientID;
        sendGameVars(msg);
        wipeBoards();   //populate boards with nothing
    }else if(msg.at("messageType")=="matchOver"){
        msg.at("client") = clientID;
        wipeBoards();   //clear the boards after each round
    }else if(msg.at("messageType")=="placeShip"){
        msg.at("client") = clientID;
        placeShip(msg);
    }else if(msg.at("messageType")=="shootShot"){
        msg.at("client") = clientID;
        shootShot(msg);
    }else if (msg.at("messageType")=="shotReturn"){ 
        shotReturned(msg);
    }else if(msg.at("messageType")=="shipDied"){ 
        //char board[10][10], int row, int col, int length, Direction dir, char newChar
        updateBoard(gameVars.shipBoard, msg.at("row"), msg.at("col"), msg.at("length"), msg.at("dir"), KILL);
    }else if (msg.at("messageType")=="killedShip"){ 
        //char board[10][10], int row, int col, int length, Direction dir, char newChar
        updateBoard(gameVars.shotBoard, msg.at("row"), msg.at("col"), msg.at("length"), msg.at("dir"), KILL);
    }
    
}

/**
 * @brief Fills the shipBoard and the shotBoard with WATER.
 */
void wipeBoards(){
    for(int row=0;row<gameVars.boardSize;row++){
        for(int col=0;col<gameVars.boardSize;col++){
            gameVars.shipBoard[row][col]=WATER;
            gameVars.shotBoard[row][col]=WATER;
        }
    }
}

/**
 * @brief The ship placement logic.
 * 
 * @param msg Contains the length of the ship, and must have the row, col, and dir be returned.
 *      json msg = {
 *          {"row", 0},
 *          {"col", 0},
 *          {"dir", HORIZONTAL},
 *          {"length", 4}
 *      }
 */
void placeShip(json &msg){ // This is an extremely bad-to-use algorithm in the long term--for more reasons than just the obvious. Try writing your own from scratch.
    int length = msg.at("length");
    for(int row=0;row<(gameVars.boardSize-length);row++){
        for(int col=0;col<(gameVars.boardSize-length);col++){
            if(gameVars.shipBoard[row][col]==WATER){
                int ship_okay = 1;
                msg.at("row") = row;
                msg.at("col") = col;
                msg.at("dir") = HORIZONTAL;
                for(int len=0;len<msg.at("length");len++){
                    if(gameVars.shipBoard[row][col]!=WATER){
                        ship_okay = 0;
                    }
                }
                if(ship_okay == 1){
                    updateBoard(gameVars.shipBoard, row, col, msg.at("length"), HORIZONTAL, SHIP);
                    return;
                }
            }
        }
    }
}

/**
 * @brief The shot taking logic.
 * 
 * @param msg Must have the row and col be returned.
 *      json msg = {
 *          {"row", 0},
 *          {"col", 0}
 *      }
 */
void shootShot(json &msg){
    for(int row=0;row<gameVars.boardSize;row++){
        for(int col=0;col<gameVars.boardSize;col++){
            if(gameVars.shotBoard[row][col]==WATER){
                msg.at("row") = row;
                msg.at("col") = col;
                updateBoard(gameVars.shotBoard, row, col, 1, NONE, SHOT);
                return;
            }
        }
    }
    msg.at("row") = 8;
    msg.at("col") = 8;
    updateBoard(gameVars.shotBoard, 8, 8, 1, NONE, SHOT);
}

/**
 * @brief shotReturned is called when there is a shot taken from either AI.
 * 
 * @param msg Contains the result of a shot at a specified row and col. Client is the AI who took the shot.
 *      json msg = {
 *          {"row", 0},
 *          {"col", 0},
 *          {"str", "HIT"},
 *          {"client", "1234"}
 *      }
 */
void shotReturned(json &msg){
    if(msg.at("client") == clientID){ // this is your AI's shot
        // do something with the message data here. 
    }else{ // this is your opponent's shot
        // do nothing
        // unless...?
    }
}
